### PR TITLE
PIN Management SDK

### PR DESCRIPTION
**Pull request by acavailhez-stripe**
## Summary
This PR introduces a `STPPinManagementService` that exposes a retrieve and update methods for PIN management (see project brief for reasoning behind arguments)

## Motivation
https://paper.dropbox.com/doc/PIN-updates--ATDftHCGwha7iEPgkRDpRxtGAg-gQnktTcgDTPZi8AFqQmlr

## Testing
New end-to-end tests

### CODE REVIEW

File: Stripe/STPPinManagementService.m [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    if (self) {
+        _keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider apiVersion:[STPAPIClient apiVersion] performsEagerFetching:NO];
+    }
+    return self;
+}
+
+- (void)retrievePin:(__unused NSString *) cardId
+         verificationId:(__unused NSString *) verificationId
+        oneTimeCode:(__unused NSString *) oneTimeCode
+         completion:(__unused STPPinCompletionBlock) completion{
```
Comment (csabol-stripe) [inline Stripe/STPPinManagementService.m]: these are used

File: Stripe/STPPinManagementService.m [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        STPAPIClient *client = [STPAPIClient apiClientWithEphemeralKey:ephemeralKey];
+        [STPAPIRequest<STPIssuingCardPin *> getWithAPIClient:client
+                                                    endpoint:endpoint
+                                                  parameters:parameters
+                                                deserializer:[STPIssuingCardPin new]
+                                                  completion:^(
+                                                               STPIssuingCardPin *details,
+                                                               __unused     NSHTTPURLResponse *response,
+                                                               NSError *error) {
+                                                      // TODO handle errors
```
Comment (csabol-stripe) [inline Stripe/STPPinManagementService.m]: resolve this TODO

File: Stripe/STPPinManagementService.m [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                                                      // Find if there were errors
+                                                      if (details.error != nil) {
+                                                          NSString* code = [details.error objectForKey:@"code"];
+                                                          if ([@"api_key_expired" isEqualToString:code]) {
+                                                              completion(nil, STPPinEphemeralKeyError, error);
+                                                          }
+                                                          else if ([@"already_redeemed" isEqualToString:code]) {
```
Comment (csabol-stripe) [inline Stripe/STPPinManagementService.m]: "official" styleguide probably hasn't been updated yet, but preference is for no newline between conditionals, i.e. this hsould be `} else if ([@"already_redeemed" isEqualToString:code]) {`

Comment (acavailhez-stripe) [reply Stripe/STPPinManagementService.m]: 👍

File: Stripe/STPPinManagementService.m (hunk 1 above)
Comment (acavailhez-stripe) [reply Stripe/STPPinManagementService.m]: ah yes, they're used within a block though and XCode got confused, do you know how to resolve that?

Comment (csabol-stripe) [reply Stripe/STPPinManagementService.m]: I just tried checking out the branch to investigate this and `STPPinManagementService.h` is missing

Comment (acavailhez-stripe) [reply Stripe/STPPinManagementService.m]: oh, hm let me check on that!

Comment (acavailhez-stripe) [reply Stripe/STPPinManagementService.m]: oh I had forgotten to add them, done now,
Does it change the `__unused` situation?

File: Stripe/STPPinManagementService.m (hunk 2 above)
Comment (acavailhez-stripe) [reply Stripe/STPPinManagementService.m]: 👍

**Comment by acavailhez-stripe** [comment]
r? @csabol-stripe 
OK Added the Update PIN and resovled comments besides the `__unused` case, where I'd love some guidance!

File: Stripe/STPPinManagementService.m (hunk 1 above)
Comment (csabol-stripe) [reply Stripe/STPPinManagementService.m]: style-nit: no space between the type and argument name. `completion` should also not have the `__unused` attribute...do you get an error when it's removed?

File: Stripe/STPPinManagementService.m [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                                                      completion(details, STPPinSuccess, nil);
+                                                  }];
+    }];
+}
+
+- (void)updatePin:(NSString *) cardId
+           newPin:(NSString *) newPin
+     verificationId:(NSString *) verificationId
+        oneTimeCode:(NSString *) oneTimeCode
+         completion:(__unused STPPinCompletionBlock) completion{
```
Comment (csabol-stripe) [inline Stripe/STPPinManagementService.m]: same comments as above

File: Stripe/STPPinManagementService.m [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                                                      }
+                                                      completion(details, STPPinSuccess, nil);
+                                                  }];
+    }];
+}
+
+- (void)updatePin:(NSString *) cardId
+           newPin:(NSString *) newPin
+     verificationId:(NSString *) verificationId
+        oneTimeCode:(NSString *) oneTimeCode
```
Comment (csabol-stripe) [inline Stripe/STPPinManagementService.m]: style-nit: remove spaces

File: Stripe/STPPinManagementService.m [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    if (self) {
+        _keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider apiVersion:[STPAPIClient apiVersion] performsEagerFetching:NO];
+    }
+    return self;
+}
+
+- (void)retrievePin:(NSString *) cardId
+         verificationId:(NSString *) verificationId
+        oneTimeCode:(NSString *) oneTimeCode
+         completion:(STPPinCompletionBlock) completion{
```
Comment (csabol-stripe) [inline Stripe/STPPinManagementService.m]: style-nit: remove spaces

File: Stripe/PublicHeaders/STPPinManagementService.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+         completion:(STPPinCompletionBlock) completion;
+
+/**
+ Updates a PIN number for a given card,
+ this call is asynchronous, implement the completion block to receive the updates
+ */
+- (void)updatePin:(NSString *) cardId
+           newPin:(NSString *) newPin
+   verificationId:(NSString *) verificationId
+      oneTimeCode:(NSString *) oneTimeCode
```
Comment (csabol-stripe) [inline Stripe/PublicHeaders/STPPinManagementService.h]: style-nit: remove spaces

File: Stripe/PublicHeaders/STPPinManagementService.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ */
+- (instancetype)initWithKeyProvider:(id<STPIssuingCardEphemeralKeyProvider>)keyProvider;
+
+/**
+ Retrieves a PIN number for a given card,
+ this call is asynchronous, implement the completion block to receive the updates
+ */
+- (void)retrievePin:(NSString *) cardId
+     verificationId:(NSString *) verificationId
+        oneTimeCode:(NSString *) oneTimeCode
```
Comment (csabol-stripe) [inline Stripe/PublicHeaders/STPPinManagementService.h]: style-nit: remove spaces

**Pull request merged**